Add MAVLink forwarding for SDK/MAVROS for Gazebo HITL

julianoes opened the pull request:
This enables MAVLink forwarding on UDP port 14540 for Gazebo HITL.

~The current state is working but has quite a bit of duplication, so I'm planning to refactor this a bit and clean it up.~
I think for the sake of not having too much churn I suggest to merge this as is, and maybe refactor it later.

With this PR Gazebo HITL actually works as described in the docs already :smile:, FYI @hamishwillee.

Also, this disables the `TCP TIME_WAIT` state on gzserver shutdown and resolves #289.

julianoes committed:
gazebo_mavlink_interface: rename addr and addr_len

This should make it clearer what the addr variables stand for.

julianoes committed:
gazebo_mavlink_interface: rename silly fd names

julianoes committed:
Add MAVLink forwarding for SDK port to Gazebo HITL

This adds forwarding to UDP port 14540 in order to enable the SDK to
connect to HITL.

The changes also separate the different sockets for different purposes,
so we don't re-use the simulator connection for QGC forwarding any
longer because that was rather confusing.

bkueng commented on src/gazebo_mavlink_interface.cpp:
```diff
@@ -549,6 +590,25 @@ void GazeboMavlinkInterface::send_mavlink_message(const mavlink_message_t *messa
+  uint8_t buffer[MAVLINK_MAX_PACKET_LEN];
+  int packetlen = mavlink_msg_to_send_buffer(buffer, message);
+
+  ssize_t len = sendto(qgc_socket_fd_, buffer, packetlen, 0, (struct sockaddr *)&remote_qgc_addr_, remote_qgc_addr_len_);
```
What if the QGC socket is not initialized?

julianoes replied:
It should be because we abort otherwise:
https://github.com/PX4/sitl_gazebo/pull/316/files/722632f16e06e68480bca36fa82a80d327e1196b..97046d68d486120c8f1b5a1d5bfddc5d896f49c1#diff-af95b4649436828244349b1766ab5c46R394

bkueng replied:
But that is conditioned on `if (serial_enabled_) {`

bkueng replied:
Has this been addressed?

bkueng commented on src/gazebo_mavlink_interface.cpp:
```diff
@@ -487,7 +523,11 @@ void GazeboMavlinkInterface::OnUpdate(const common::UpdateInfo&  /*_info*/) {
   SendSensorMessages();
 
-  pollForMAVLinkMessages();
+  if (!serial_enabled_) {
```
I'd rather invert the logic: `if (serial_enabled_) {`

julianoes replied:
Fixed in c19fc4b.

bkueng reviewed (commented):
Thanks.
I somewhat dislike the duplicated code for the SDK and QGC sockets. Wrapping them into an array would avoid that.

julianoes committed:
gazebo_mavlink_interface: invert if logic

julianoes committed:
gazebo_mavlink_interface: close on SIGINT

This adds a connection to the SIGINT event, so that we can try to
close() all sockets before we get destroyed. We also call close() in the
destructor in case SIGINT is not called, however, that case is probably
unlikely.

julianoes committed:
gazebo_mavlink_interface: disable TCP TIME_WAIT

From what I understand, we don't need the TIME_WAIT state of the TCP
socket after closing the connection.

bkueng commented on include/gazebo_mavlink_interface.h:
```diff
@@ -357,5 +383,7 @@ class GazeboMavlinkInterface : public ModelPlugin {
   bool hil_mode_;
   bool hil_state_level_;
+
+  std::atomic<bool> gotSigInt_ {false};
```
```suggestion
  std::atomic<bool> got_sig_int_ {false};
```

Although I see we already have inconsistencies.

julianoes committed:
gazebo_mavlink_interface: check fds

We don't need this but it might be needed after a future refactoring.

julianoes committed:
gazebo_mavlink_interface: fix HITL without serial

This is an attempt to unbreak HITL over something other than serial.
Unfortunately, I don't have the setup to test this but it should be more
correct than the previous state.
